cmd/healthcheck-handler.go

```diff
@@ -40,19 +42,38 @@ func ReadinessCheckHandler(w http.ResponseWriter, r *http.Request) {
 	writeResponse(w, http.StatusOK, nil, mimeNone)
 }
 
-// LivenessCheckHandler -- checks if server can ListBuckets internally. If not, server is
-// considered to have failed and needs to be restarted.
+// LivenessCheckHandler -- checks if server can reach its disks internally.
+// If not, server is considered to have failed and needs to be restarted.
 // Liveness probes are used to detect situations where application (minio)
 // has gone into a state where it can not recover except by being restarted.
 func LivenessCheckHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(r, w, "LivenessCheckHandler")
+
 	objLayer := newObjectLayerFn()
 	// Service not initialized yet
 	if objLayer == nil {
 		writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
 		return
 	}
-	// List buckets is unsuccessful, means server is having issues, send 503 service unavailable
-	if _, err := objLayer.ListBuckets(context.Background()); err != nil {
+	var totalLocalDisks int
+	var erroredDisks int
+	for _, endpoint := range globalEndpoints {
+		// Check only if local disks are accessible, we do not have
+		// to reach to rest of the other servers in a distributed setup.
+		if endpoint.IsLocal {
+			totalLocalDisks++
+			// Attempt a stat to backend, any error resulting
+			// from this Stat() operation is considered as backend
+			// is not available, count them as errors.
+			if _, err := os.Stat(endpoint.Path); err != nil {
+				logger.LogIf(ctx, err)
+				erroredDisks++
+			}
+		}
+	}
+	// If all exported local disks have errored, we simply let kubernetes
+	// take us down.
+	if totalLocalDisks == erroredDisks {
```

Comment:
If all the local disks are unavailable then kubernetes will keep restarting the container endlessly.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

> liveness probes could catch a deadlock, where an application is running, but unable to make progress. Restarting a Container in such a state can help to make the application more available despite bugs.

From this document the purpose of this endpoint seems to be to restart hung processes (ex. deadlocked processes) It is hard to think of a situation where minio would hang (in which case we have to detect such cases and return error in this endpoint) If we can think of ways to detect situations where minio would hang then we should test for such a situation in this handler and return error.

i.e the logic should be:
```
if(minio_is_hung()) {
    writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
    return
}
writeResponse(w, http.StatusOK, nil, mimeNone)
```

minio_hung() -> should have code to detect if minio is hung/deadlocked

Reply:
I don't think we can think of a case where minio would be hung there is no such instrumentation built in right now neither we can foresee such thing anyways since we have timeouts on most situations.

Even previously the primary purpose of this code was that we can restart the process if the PV is somehow unavailable as it shouldn't happen. Upon unavailable disks restarting the POD might fix the problem. So PR is correct in its own way.

Reply:
> I don't think we can think of a case where minio would be hung there is no such instrumentation built in right now neither we can foresee such thing anyways since we have timeouts on most situations.

The kubernetes doc and the above comment indicates that we should just be replying StatusOK all the time (since we don't have deadlock detection) Approving the PR though so as not to block it.

Reply:
They never mention that liveness is only meant for deadlocks. For k8s it is all about pods and services it runs a service which experiences a broken state in case of Minio is having the unexpected disk situation, a situation where PV is not available is as good a situation to let k8s know our pod is experiencing problems.